src/metatensor/models/utils/data/combine_dataloaders.py

```diff
@@ -1,57 +1,38 @@
-import itertools
 from typing import List
 
 import numpy as np
 import torch
 
 
-class CombinedIterableDataset(torch.utils.data.IterableDataset):
-    """
-    Combines multiple dataloaders into a single iterable dataset.
-    This is useful for combining multiple dataloaders into a single
-    dataloader. The new dataloader can be shuffled or not.
-
-    :param dataloaders: list of dataloaders to combine
-    :param shuffle: whether to shuffle the combined dataloader
-
-    :return: combined dataloader
-    """
-
-    def __init__(self, dataloaders, shuffle):
+class CombinedDataLoader:
+    def __init__(self, dataloaders: List[torch.utils.data.DataLoader], shuffle: bool):
         self.dataloaders = dataloaders
         self.shuffle = shuffle
 
         # Create the indices:
-        indices = [
-            (i, dl_idx)
-            for dl_idx, dl in enumerate(self.dataloaders)
-            for i in range(len(dl))
-        ]
+        self.indices = list(range(len(self)))
 
         # Shuffle the indices if requested
         if self.shuffle:
-            np.random.shuffle(indices)
-
-        self.indices = indices
+            np.random.shuffle(self.indices)
 
-    def __iter__(self):
-        for idx, dataloader_idx in self.indices:
-            yield next(itertools.islice(self.dataloaders[dataloader_idx], idx, None))
+        self.reset()
 
-    def __len__(self):
-        return len(self.indices)
+    def reset(self):
+        self.current_index = 0
+        self.full_list = [batch for dl in self.dataloaders for batch in dl]
 
+    def __iter__(self):
+        return self
 
-def combine_dataloaders(
-    dataloaders: List[torch.utils.data.DataLoader], shuffle: bool = True
-):
-    """
-    Combines multiple dataloaders into a single dataloader.
+    def __next__(self):
+        if self.current_index >= len(self.indices):
+            self.reset()  # Reset the index for the next iteration
+            raise StopIteration
 
-    :param dataloaders: list of dataloaders to combine
-    :param shuffle: whether to shuffle the combined dataloader
+        idx = self.indices[self.current_index]
+        self.current_index += 1
+        return self.full_list[idx]
 
-    :return: combined dataloader
-    """
-    combined_dataset = CombinedIterableDataset(dataloaders, shuffle)
-    return torch.utils.data.DataLoader(combined_dataset, batch_size=None)
+    def __len__(self):
+        return sum(len(dl) for dl in self.dataloaders)
```

Comment:
this will change the behavior from before? Isn't this weird that if you ask for the `len` of a combined dataloader and you get the summed length of all elements? Usually a length of a list of arrays returns the number of arrays and not the sum the length of arrays...

Reply:
Yes, it's just a matter of definitions. In this case, I think it should return the number of batches that the dataloader returns in one epoch to match the behavior of torch dataloaders. The number of batches that it returns is the sum of the number of batches in the individual dataloaders

Reply:
I am not sure. But then we have to comment this at least.

Reply:
Added the comment

Reply:
I agree that a sum is the behavior we want (the length of a torch dataset is always the total number of elements inside you can index) if we decide to hide the fact that this is a combined dataset, i.e. if the user should use it as 

```py
for batch in dataset:
   # do stuff
```

However, my understanding was that the batches would not mix stuff from different datasets, is this right? if so, what happens when combining a dataset of size 7 and one of size 11 and trying to use a batch size of 10?

Reply:
This is a combined dataloader, not dataset. It takes many dataloaders and returns the individual batches (shuffled or not, depending on what the caller wants)

Reply:
right, makes sense then!

Reply:
right, and then the length of the combined dataloader is how many batches it will produce. Makes sense.